### PR TITLE
raftstore: Implement coprocessor observer post_exec_admin(query)

### PR DESCRIPTION
**Pull request by CalvinNeo**
### What is changed and how it works?

Issue Number: ref #12849

What's Changed:

According to #12849, we need to support some new observers, including post_exec_admin(query) which is implemented in this PR.
```commit-message
Support new observers post_exec_admin(query).
```

### Related changes

- PR to update `pingcap/docs`/`pingcap/docs-cn`:
- Need to cherry-pick to the release branch

### Check List 

Tests 

- Unit test
- Integration test
- Manual test (add detailed scripts or steps below)
- No code

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM
- Breaking backward compatibility

### Release note 

```release-note
None
```

### CODE REVIEW

File: components/raftstore/src/store/fsm/apply.rs [hunk 1]
```diff
@@ -1323,6 +1346,10 @@ where
             }
         }
 
+        if should_write {
+            info!("persist data and apply state"; "region_id" => self.region_id(), "peer_id" => self.id(), "state" => ?self.apply_state);
+            ctx.commit(self);
```
Comment (CalvinNeo) [inline components/raftstore/src/store/fsm/apply.rs:1351]: Should we move this into an outer function?

**Comment by CalvinNeo** [comment]
> Add some test for changes in apply.rs

Will do that after pre_exec_query(admin) are merged.

File: components/raftstore/src/store/fsm/apply.rs [hunk 2]
```diff
@@ -5491,6 +5550,102 @@ mod tests {
         });
     }
 
+    #[test]
+    fn test_exec_observer() {
```
Comment (tonyxuqqi) [inline components/raftstore/src/store/fsm/apply.rs:5554]: add some comments for the test steps and the logic to verify.

**Review by tonyxuqqi — APPROVED** [review]

File: components/raftstore/src/store/fsm/apply.rs (hunk 2 above)
Comment (CalvinNeo) [reply components/raftstore/src/store/fsm/apply.rs:5554]: Will do after pre_exec is merged

File: components/cdc/src/delegate.rs [hunk 3]
```diff
@@ -446,6 +446,7 @@ impl Delegate {
         for cmd in batch.into_iter(self.region_id) {
             let Cmd {
                 index,
+                term: _,
```
Comment (BusyJay) [inline components/cdc/src/delegate.rs:449]: What's this?

File: components/raftstore/src/coprocessor/dispatcher.rs [hunk 4]
```diff
@@ -416,6 +416,33 @@ impl<E: KvEngine> CoprocessorHost<E> {
         }
     }
 
+    pub fn post_exec(
```
Comment (BusyJay) [inline components/raftstore/src/coprocessor/dispatcher.rs:419]: Why not use `post_apply`?

File: components/cdc/src/delegate.rs (hunk 3 above)
Comment (CalvinNeo) [reply components/cdc/src/delegate.rs:449]: We don't use `term` here.

File: components/raftstore/src/coprocessor/dispatcher.rs (hunk 4 above)
Comment (CalvinNeo) [reply components/raftstore/src/coprocessor/dispatcher.rs:419]: The existing `post_apply` observer is not enough. Only `pre_exec` can observe more necessary execution information we need, at a earlier time than `post_apply`.

Comment (BusyJay) [reply components/raftstore/src/coprocessor/dispatcher.rs:419]: Please add some comment to describe the difference.

Comment (CalvinNeo) [reply components/raftstore/src/coprocessor/dispatcher.rs:423]: addressed

File: components/raftstore/src/store/fsm/apply.rs (hunk 1 above)
Comment (BusyJay) [reply components/raftstore/src/store/fsm/apply.rs:1373]: Yes, otherwise the response won't be returned for the last command.

**Review by BusyJay — COMMENTED** [review]
I think we need to bench the performance impact after adding two observers. sysbench should be enough.

**Comment by CalvinNeo** [comment]
> I think we need to bench the performance impact after adding two observers. sysbench should be enough.
It is tested that the performance is quite close to each other.
VHH72SaHAz

**Review by BusyJay — APPROVED** [review]

**Pull request merged**